### This Document contains the links to publicly available data sources of annotations and CNVs

## CNVs

# GnomAD SVs (BED-file)
wget https://storage.googleapis.com/gnomad-public/papers/2019-sv/gnomad_v2_sv.sites.bed.gz

# Variants called by the Eichler group (BED-file)
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/hgsv_sv_discovery/working/20181025_EEE_SV-Pop_1/VariantCalls_EEE_SV-Pop_1/EEE_SV-Pop_1.ALL.sites.20181204.bed.gz

## Annotations

# GnomAD Genes
wget https://storage.googleapis.com/gnomad-public/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz

# DDG2P Genes
wget http://www.ebi.ac.uk/gene2phenotype/downloads/DDG2P.csv.gz

# Genes annotated with predicited Hapliinsufficiency
wget https://decipher.sanger.ac.uk/files/downloads/HI_Predictions_Version3.bed.gz

# FANTOM5 Enhancer
wget http://fantom.gsc.riken.jp/5/datafiles/latest/extra/Enhancers/human_permissive_enhancers_phase_1_and_2.bed.gz

# TAD boundaries called by Dali et al.
wget https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5389712/bin/gkx145_Supp.zip

# Human Accelerated Regions (HARs) (Excel-file)
wget https://www.cell.com/cms/10.1016/j.cell.2016.08.071/attachment/9e010545-053a-4166-9b0d-7cb8c8f90ca7/mmc2.xlsx

# Chrom HMM Predictions
wget http://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeBroadHmm/wgEncodeBroadHmmH1hescHMM.bed.gz

# Segmental Duplications
wget http://humanparalogy.gs.washington.edu/build37/data/GRCh37GenomicSuperDup.tab

# Crossover locations
wget https://science.sciencemag.org/highwire/filestream/721792/field_highwire_adjunct_files/5/aau1043_DataS4.gz

## Conservation Scores
# I excluded the conservation scores from any tutorial due to the file size

# Phastcon
# This file is over 5 GB.
# http://hgdownload.cse.ucsc.edu/goldenpath/hg19/phastCons100way/hg19.100way.phastCons.bw

# PhyloP
# This file is over 8 GB
# http://hgdownload.cse.ucsc.edu/goldenpath/hg19/phyloP100way/hg19.100way.phyloP100way.bw
